### PR TITLE
Sticky headers layout manager

### PR DESCRIPTION
**Pull request by goncalossilva**
This PR switches the sticky headers approach from `ItemDecoration` to an extension of `LinearLayoutManager`.

Specifically, it fixes the following problems:
- Sticky header being above everything (eg. the scrollbar) or below everything (eg. the list itself)
- Not using `Recycler` to obtain a new view, hence not having its `LayoutParams` properly set and being unable to return it to the pool in some scenarios

It has a couple more advantages:
- Simplifies workflow greatly
- Performance gains (since the view is added to the `RecyclerView` itself, layout / invalidation are part of the regular work. For a total of 280 runs, it takes an average of 0.175ms with 60% variation versus 0.250ms with 180% variation of the previous fastest implementation, due to extra allocations)

And as usual, a couple of disadvantages:
- A lot of boilerplate around `LinearLayoutManager`'s scroll methods. Namely, we have to detach and re-attach (quick op) when they're called, to avoid breaking its calculations — the problem is that the last child (ie. the sticky header) is positioned at the top, which for these methods doesn't make sense)
- Inability to run in the animation cycle, as it gets executed in the layout cycle (I don't see any cases for this right now, so it's not really a disadvantage — but it could be in the future)

### CODE REVIEW

File: StickyHeaders/src/main/java/io/doist/recyclerviewext/sticky_headers/StickyHeadersLinearLayoutManager.java [hunk 1]
```diff
@@ -15,243 +13,339 @@
...
 
-                    // Ensure sticky header is created, if absent, or bound, if dirty or the position changed.
+                    // Ensure sticky header is created, if absent, or bound, if being laid out or the position changed.
                     if (mStickyHeader == null) {
-                        createStickyHeader(parent, headerPos);
+                        createStickyHeader(recycler, headerPos);
                     }
-                    if (mDirty || mStickyHeader.getLayoutPosition() != headerPos) {
-                        bindStickyHeader(parent, headerPos);
+                    if(layout || getPosition(mStickyHeader) != headerPos) {
```
Comment (rastislavvasko) [inline StickyHeaders/src/main/java/io/doist/recyclerviewext/sticky_headers/StickyHeadersLinearLayoutManager.java]: Missing space after `if`. :)

**Comment by rastislavvasko** [comment]
Looks good to me! :+1: Couldn't find any issue and the disadvantages seem minor.

File: StickyHeaders/src/main/java/io/doist/recyclerviewext/sticky_headers/StickyHeadersLinearLayoutManager.java [hunk 2]
```diff
@@ -15,243 +13,336 @@
...
      */
-    protected void scrapStickyHeader(final RecyclerView parent) {
-        onScrapStickyHeader(mStickyHeader, parent);
+    protected void scrapStickyHeader(RecyclerView.Recycler recycler) {
+        // Revert translation values.
+        mStickyHeader.setTranslationX(0);
+        mStickyHeader.setTranslationY(0);
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            mStickyHeader.setTranslationZ(0);
```
Comment (rastislavvasko) [inline StickyHeaders/src/main/java/io/doist/recyclerviewext/sticky_headers/StickyHeadersLinearLayoutManager.java]: Do we need to reset `translationZ` as we aren't setting it anywhere?

Comment (goncalossilva) [inline StickyHeaders/src/main/java/io/doist/recyclerviewext/sticky_headers/StickyHeadersLinearLayoutManager.java]: Good catch — left-over from previous implementation :smile: Removed, thanks!

**Pull request merged**